Fixes sphzone, sphlayer, cyzone, cylayer empty selections

orioncohen opened the pull request:
Fixes #2915 

Changes made in this Pull Request:
 - created a new decorator to check for empty sub-expressions in sphzone, sphlayer, cyzone, and cylayer.
 - added testing to confirm the fix
 - updated CHANGELOG
 - updated AUTHORS


PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

orioncohen commented:
Could someone help me understand what codecov is doing and how I can fix the error it is throwing?

lilyminium reviewed (changes requested):
Thanks for looking at this, @orioncohen. LGTM. As this is your first contribution to MDAnalysis, please add yourself to AUTHORS -- please also add your username to the top of CHANGELOG :-)

lilyminium commented:
Thanks @orioncohen. Looking more closely, though, the issue actually mentions cyzone and (in a comment lower down) sphlayer. As a more general fix could you look at those too? In fact it may be more elegant to just modify the `return_empty_on_apply` decorator or make a new decorator, to follow the Don't Repeat Yourself philosophy of Python.

orioncohen commented:
Hi @lilyminium. I've added a fix for cyzone, cylayer, and sphlayer too. I tried creating a new decorator to test if the selection is empty. The problem is that we need to check the contents of `sel` after calling`sel = self.sel.apply(group)`, so the decorator needs to reference an internal method of the object. The best workaround I can think of is to add a decorator method to the Selection base class but that appears to be a meta-class. I ended up going with the naive solution, which required adding 4 more lines of code to fix cyzone, cylayer, and sphlayer.

lilyminium commented:
> so the decorator needs to reference an internal method of the object. The best workaround I can think of is to add a decorator method to the Selection base class but that appears to be a meta-class. 

Your solution does fix the problem, but I'm surprised you can't just decorate the `apply()` methods of the relevant Selection classes the same way that `return_empty_on_apply` has been used.

pep8speaks commented:
Hello @orioncohen! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`package/MDAnalysis/core/selection.py`](https://github.com/MDAnalysis/mdanalysis/blob/96b63e3e47c9508753e06024f7928e1d4803a4a6/package/MDAnalysis/core/selection.py):

> [Line 220:1](https://github.com/MDAnalysis/mdanalysis/blob/96b63e3e47c9508753e06024f7928e1d4803a4a6/package/MDAnalysis/core/selection.py#L220): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

orioncohen commented:
I should have taken a closer look when I was less sleepy! I implemented this with a decorator as you suggested. I added a new decorator rather than modifying `return_empty_on_apply` because not all of the selection methods call `self.sel.apply(group)`. I also moved around some testing to be a bit more consistent.

lilyminium commented on package/MDAnalysis/core/selection.py:
```diff
@@ -348,7 +363,7 @@ def __init__(self, parser, tokens):
         self.sel = parser.parse_expression(self.precedence)
 
-    @return_empty_on_apply
+    @return_empty_on_empty_selection
```
Rather than replacing `return_empty_on_apply`, you can usually stack decorators like

```
@return_empty_on_apply
@return_empty_on_empty_selection
```

orioncohen replied:
Implemented this too!

lilyminium commented on package/MDAnalysis/core/selection.py:
```diff
@@ -217,6 +217,21 @@ def apply(self, group):
+    """
+    @functools.wraps(func)
+    def apply(self, group):
+        if len(group) == 0:
```
If you stack decorators as I commented above, you could avoid replicating the functionality of `return_empty_on_apply`.

lilyminium commented on package/MDAnalysis/core/selection.py:
```diff
@@ -217,6 +217,21 @@ def apply(self, group):
+    def apply(self, group):
+        if len(group) == 0:
+            return group
+        if len(self.sel.apply(group)) == 0:
+            return group[[]]
+        return func(self, group)
```
@richardjgowers will know efficient this is and whether the following suggestion is a good idea, but calling `self.sel.apply(group)` twice (once in the decorator, once in the function it's decorating) seems a little wasteful. Could you do `sel = self.sel.apply(group)` and pass `sel` in as an extra keyword argument? This would involve modifying the function signature of the relevant `apply`s, which might be a bad idea.

orioncohen replied:
I could definitely just pass `sel` in as an extra keyword. I don't like that it would make the function signature of `apply` inconsistent between classes though. I could also go back to implementing it in the functions, but that's very repetitive. This seems like largely a stylistic choice, so just let me know how I should proceed!

richardjgowers replied:
Yeah I think Lily is right that this is neater but less efficient.  I'd say duplicate the code a few times

orioncohen replied:
Done. I let me know if there is anything else I should do before this gets merged.

lilyminium commented on testsuite/MDAnalysisTests/core/test_atomselections.py:
```diff
@@ -209,6 +209,8 @@ def test_sphzone(self, universe, selstr):
     def test_cylayer(self, universe, selstr):
         sel = universe.select_atoms(selstr)
         assert_equal(len(sel), 88)
+        empty = universe.select_atoms('cylayer 4.0 6.0 10 -10 name NOT_A_NAME')
```
Thanks, these are thorough tests -- could you make each of these separate though? That way each function tests one specific functionality.

orioncohen replied:
Yep! just fixed this!

lilyminium reviewed (changes requested):
Thanks @orioncohen, the decorator is much nicer than before! I just have some design comments. @richardjgowers you know more than I do about selections, so please weigh in if I've said anything dumb.

lilyminium commented on package/CHANGELOG:
```diff
@@ -17,11 +17,12 @@ The rules for this file:
   * 2.0.0
 
 Fixes
+  * Fixed 'sphzone' behavior to be consistent with 'around' (Issue #2915)
```
Could you now update this CHANGELOG message to be more informative?

orioncohen replied:
Done! Thanks for the help Lily! (loved your PyCon talk btw)

lilyminium replied:
Thank you @oriencohen! :-)

lilyminium reviewed (changes requested):
Other than the CHANGELOG message, LGTM!
